src/state/actions/workspace.js

```diff
@@ -68,3 +73,40 @@ export function toggleWorkspaceExposeMode() {
     type: ActionTypes.TOGGLE_WORKSPACE_EXPOSE_MODE,
   };
 }
+
+/**
+ * importWorkspace - action creator
+ */
+export function importWorkspace(stateExport) {
+  return (dispatch, getState) => {
+    dispatch(importConfig(stateExport.config));
+    const { viewers } = stateExport || {};
+    const imWins = values(stateExport.windows);
+
+    /* re-use existing windows */
+    const exIds = values(getState().windows).map((exWin) => {
```

Comment:
Can you explain a little bit of the context around this? This seems like an implicit state mutation ("reusing" property values). What were some of the blocks you ran into here that this addresses?

Reply:
(Immutability is extremely important for maintainability and debugging.)

Reply:
I wonder if this is related to state mutation, but I noticed if I copied some state, added some windows, and then tried to import the previous state, I got a javascript error:

![2019-03-11 15-02-13 2019-03-11 15_07_56](https://user-images.githubusercontent.com/111218/54161495-88184180-440f-11e9-8e46-04ef93db290a.gif)

Reply:
> Can you explain a little bit of the context around this? This seems like an implicit state mutation ("reusing" property values). What were some of the things you ran into here that required a reuse pattern?

The idea is the following: If the workspace, that is already opened when starting the import process, already contains windows e. g. viewer instances, we can re-use these windows by dispatching an `UPDATE_WINDOW` action to optimize the performance of the import process. All surplus windows will be closed by dispatching a `REMOVE_WINDOW` action. If, on the other hand, the number of windows to be imported is higher than the number of the existing ones, we need to create new instances by dispatching an `ADD_WINDOW` action. But no state manipulation is done directly here.